azext_iot/common/base_discovery.py

```diff
@@ -0,0 +1,359 @@
+                    resource_name=resource_name,
+                    resource_group_name=rg,
+                    key_name=policy_name
+                )
+            )
+
+        policy_list = self.get_policies(resource_name=resource_name, rg=rg)
+
+        for policy in policy_list:
+            rights_set = set(policy.rights.split(", "))
+            if self.necessary_rights_set.issubset(rights_set):
+                logger.info(
+                    "Using policy '%s' for %s interaction.", policy.key_name, self.resource_type
+                )
+                return policy
+
+        raise CLIError(
+            POLICY_ERROR_TEMPLATE.format(
+                self.resource_type,
+                resource_name,
+                self.sub_id,
+                _format_policy_set(self.necessary_rights_set)
+            )
+        )
+
+    @classmethod
+    @abstractmethod
+    def get_target_by_cstring(cls, connection_string):
+        """Returns target inforation needed from a connection string."""
+        pass
+
+    def get_target(
+        self, resource_name: str, resource_group_name: str = None, **kwargs
+    ) -> Dict[str, str]:
+        """
+        Returns a dictionary of the given resource's connection string parts to be used
+        by the extension.
+
+        This function finds the target resource and builds up a dictionary of connection
+        string parts needed for IoT extension operation. In future iteration we will
+        return a 'Target' object rather than dict but that will be better served aligning
+        with vNext pattern for Iot Hub/DPS.
+
+        If the resource group is not provided, will look through all resources within the
+        subscription and return first match. This functionality will only work for
+        resource types that require unique names within the subscription.
+
+        If the policy name is not provided, will look through all policies for the given
+        resource and return the first usable policy (the first policy that the IoT
+        extension can use).
+
+        Raises CLIError if no resource is found.
+
+        :param resource_name: Resource Name
+        :type resource_name: str
+        :param rg: Resource Group
+        :type rg: str
+
+        :keyword str login: Connection string for the target resource
+        :keyword str key_type: Key type to use in connection string construction
+        :keyword auth_type: Authentication Type for the Dataplane
+        :paramtype auth_type: AuthenticationTypeDataplane
+        :keyword str policy_name: Policy name to use
+
+        :return: Resource
+        :rtype: dict representing self.resource_type
+        """
+        cstring = kwargs.get("login")
+        if cstring:
+            return self.get_target_by_cstring(connection_string=cstring)
+
+        resource_group_name = resource_group_name or kwargs.get("rg")
+        target = self.find_resource(resource_name=resource_name, rg=resource_group_name)
```

Comment:
It would be more apt to not call the result of `self.find_resource()` target as the term `target` has specific meaning in discovery context. Also looking at later interactions, it is strange to call say `self._build_target(resource=target, ...)` to return a target dictionary while passing in "target" for the resource parameter.

Reply:
I'm Ok to cover this in a subsequent PR.